### PR TITLE
InteriorBasis.point_source

### LINKED ISSUE #629: point source
I commented on the old issue #561 

> Is the point-source vector just the transpose of the probe matrix #601?

but will open a new issue to investigate this.  It might be nice feature to add to the library, or at least demonstrate in an example.

The thinking was that interpolating a function *u* at a point **p** is like (*u*, &delta;(**x** &minus; **p**)) whereas a point-source right-hand side vector is like (&delta;(**x** &minus; **p**), *v*).

### PR DESCRIPTION
**Pull request by gdmcbain**
Fix #629 and exemplify.

ex38

This won't work for `MeshLine` until #628 is resolved, but is fine for `MeshTri`, &c.

Should probably compare the finite element solution with Green's function, as discussed in the docstring.  For a test, I'd suggest one dimension, pending #628.

### CODE REVIEW

File: skfem/assembly/basis/interior_basis.py [hunk 1]
```diff
@@ -154,6 +154,10 @@ def probes(self, x: ndarray) -> coo_matrix:
             shape=(x.shape[1], self.N),
         )
 
+    def point_source(self, x: ndarray) -> ndarray:
+        """Return right-hand side vector for unit source at `x`"""
```
Comment (kinnala) [inline skfem/assembly/basis/interior_basis.py:158]: Is this docstring true for other than P1 Lagrange?

Comment (gdmcbain) [reply skfem/assembly/basis/interior_basis.py:158]: Hmm, I think so, in the Galerkin sense. Initially I reread Jan's point-source function in #561 and was reminded of what we'd done for the probe for Nico. Then i thought that at the continous variational level, interpolation is the inner product of a finite element function with a Dirac delta and a point source is the inner product of a Dirac delta with a test function. Those are adjoints and on discretization they're transposes. That line of thoight doesn't make any reference to the txpe of basis.

Similarly for `skfem.models.poisson.unit_load` and the functional that integrates, replacing the Dirac delta with the constant 1.

Comment (gdmcbain) [reply skfem/assembly/basis/interior_basis.py:158]: I don't have a reference for this. I couldn't find one in a quick look. I remember that for years users were asking about point sources on other finite element msiling lists. Somtimes the discussion suggested mollification: convolving the point source with a Gaussian kernel, but I didn't like that so much as it still didn't seem well suited to the quadrature.  Then FreeFEM added one recently; I refrained from looking at  their implementation as it's GPL.

Comment (gdmcbain) [reply skfem/assembly/basis/interior_basis.py:158]: Possible hole in the argument: the argument shows that the transpose property holds but it only follows from that that the function returns the Galerkin point source vector if the probes function works correctly. I didn't look into whether that works for all kinds of bases. Does it? I assumed that it did.

Comment (gdmcbain) [reply skfem/assembly/basis/interior_basis.py:158]: A couple of brief references to point sources in some of the older FEM books on my self:
* Segerlind, L. J. (1976). *Applied Finite Element Analysis.* New York: Wiley, &sect;8.6 &lsquo;Point sources&rsquo;
* Hughes, T. J. R. (2000). *The Finite Element Method.* Mineola, New York: Dover, exercise 3.11.2(*b*)

Unfortunately Hughes doesn't give the solution to the exercise (delta function loading on a one-dimensional, quadratic, three-node element), but I think it's looking the similar to my pair of [linear two-node elements](https://github.com/kinnala/scikit-fem/issues/629#issuecomment-823701775).

Segerlind notes:

> The integral of a function multiplied by an impulse function, however, is simply the function evaluated at <i>X</i><sub>0</sub> and <i>Y</i><sub>0</sub>.

which is elementary and the subsequent example is only for `ElementTriP1`.

I'm yet to turn up anything on nonlagrangian elements.

**Comment by gdmcbain** [comment]
```python
from pathlib import Path

import numpy as np

import skfem
from skfem.models import laplace
from skfem.visuals.matplotlib import plot, savefig

mesh = skfem.MeshLine(np.linspace(0, 1, 2 ** 1 + 1))
basis = skfem.InteriorBasis(mesh, skfem.ElementLineHermite())

lap = skfem.asm(laplace, basis)
print(lap.toarray())

a = 3 / 4
rhs = basis.point_source(np.array([a]))
print("RHS:", rhs)
condensed = skfem.condense(lap, rhs, D=basis.find_dofs())
print(condensed[0].toarray())
print(condensed[1])

solution = skfem.solve(*condensed)
print(solution)

ax = plot(basis, solution, Nrefs=3)
ax.set_title(f"{type(basis.elem)}")
x = np.linspace(mesh.p[0].min(), mesh.p[0].max())
ax.plot(x, np.stack([(1 - a) * x, (1 - x) * a]).min(0), linestyle="--", label="exact")
ax.plot(
    mesh.p[0],
    solution[basis.nodal_dofs[0]],
    marker="o",
    linestyle="None",
    label="nodal DoFs",
)
ax.legend()
savefig(Path(__file__).with_suffix(".png"))
```

mwe

I think that that's doing the right thing.  The `plot` looks a bit funny though. I copied from

https://github.com/kinnala/scikit-fem/blob/13b5d697d41d96417abb3288fa524bb4f02290b9/docs/examples/ex34.py#L47

`ElementLineMini` looks O. K. too.

mwe

**Comment by gdmcbain** [comment]
I mean Hermite elements are not a good choice for this problem since they're C<sup>1</sup> and the solution isn't, but they're trying:

mwe

Actually
```python
condensed = skfem.condense(lap, rhs, D=basis.find_dofs())
```
isn't the right boundary condition for Hermite either, that's for a built-in Euler&ndash;Bernoulli beam, isn't it.

**Comment by gdmcbain** [comment]
Perhaps I'll look at the elastic plate with `ElementTriMorley` too, adapting ex18.
```python
from skfem import *
from skfem.helpers import ddot, dd

import numpy as np


mesh = MeshTri.init_circle(4)
element = ElementTriMorley()
mapping = MappingAffine(mesh)
ib = InteriorBasis(mesh, element, mapping, 2)


@BilinearForm
def biharmonic(u, v, w):
    return ddot(dd(u), dd(v))


stokes = asm(biharmonic, ib)
psi = solve(*condense(stokes, ib.point_source(np.array([0.2, 0.3])), D=ib.find_dofs()))


if __name__ == "__main__":
    from os.path import splitext
    from sys import argv
    from skfem.visuals.matplotlib import draw, show
    from matplotlib.tri import Triangulation

    M, Psi = ib.refinterp(psi, 3)

    ax = draw(mesh)
    ax.tricontour(Triangulation(*M.p, M.t.T), Psi)
    name = splitext(argv[0])[0]
    show()
```

ex18_point-source

**Comment by kinnala** [comment]
Looks good. Ready for merge or still a draft?

**Comment by kinnala** [comment]
I've actually used this kind of load in some of the articles on Kirchhoff plates and always hacked it together in nonreusable way. This would've really simplified that. Thanks!

**Comment by gdmcbain** [comment]
> Looks good. Ready for merge or still a draft?

I should add a test.

I'll add a comparison with the Green's function to the example later.

I assume the Green's function for the Kirchhoff plate is known too.  Have you got a reference to hand?  Actually that might make a nicer example than the membrane, showing right away that it works for other than Lagrange elements.

**Comment by kinnala** [comment]
> I assume the Green's function for the Kirchhoff plate is known too. Have you got a reference to hand?

I think there is but I have no reference at hand.

On the other hand, there is an infinite series solution available for a square plate and a point load. The series is not converging very quickly but it's something if you truncate it to a sum. See Section 5.1 (eq. 5.13) in this preprint: https://arxiv.org/pdf/1707.08396.pdf

**Comment by gdmcbain** [comment]
Hang on, something goes wrong if I change from `MeshLine(np.linspace(0, 1, ...))` to `MeshLine().refined(...)`.
mwe

**Comment by gdmcbain** [comment]
```
>>> m = fe.MeshLine().refined()
>>> fe.InteriorBasis(m, fe.ElementLineP1()).probes(np.array([[3/4]])).toarray()
array([[-0.5,  0. ,  1.5]])
>>> fe.InteriorBasis(fe.MeshLine(m.p[0]), fe.ElementLineP1()).probes(np.array([[3/4]])).toarray()
array([[0. , 0.5, 0.5]])
```

**Comment by gdmcbain** [comment]
Is the second element in `MeshLine().refined()` everted?
```python
>>> m.t
array([[0, 2],
       [2, 1]])
```

**Comment by gdmcbain** [comment]
It looks like `MeshLine.element_finder` #417 #628 again.
```
>>> m = MeshLine().refined()
>>> m.element_finder()(np.array([.7]))
array([0])
>>> MeshLine(m.p[0]).element_finder()(np.array([.7]))
array([1])
```

**Comment by kinnala** [comment]
> It looks like `MeshLine.element_finder` #417 #628 again.
> 
> ```
> >>> m = MeshLine().refined()
> >>> m.element_finder()(np.array([.7]))
> array([0])
> >>> MeshLine(m.p[0]).element_finder()(np.array([.7]))
> array([1])
> ```

There is `Mesh.sort_t` flag which defaults to `False`. `MeshTri1` has `sort_t=True` which is explained in the docstring of `Mesh`.

**Comment by gdmcbain** [comment]
Let's take that up in #633.

**Comment by kinnala** [comment]
Let me add, `sort_t=False` is something I'd prefer to have everywhere but it wasn't easy to drop it from `MeshTri1` because other code assumes it. So maybe sorting in `element_finder` is sufficient?

**Comment by gdmcbain** [comment]
It remains to include the example in `tests/test_examples` and before that to put something testable in the example, e.g. comparison of the solution with the analytic Green's function (once I track down a reference for that&mdash;basically it's the usual logarithmic source minus an image sink on the other side of the circumference).

**Comment by gdmcbain** [comment]
I still haven't been able to find a better reference for the Green's function (Sadybekov et al just quote the result without reference or derivation), but this is it anyway.

**Comment by gdmcbain** [comment]
The PNG [above](https://github.com/kinnala/scikit-fem/pull/632#issue-619872748) will do for the Gallery.

**Comment by kinnala** [comment]
> The PNG [above](https://github.com/kinnala/scikit-fem/pull/632#issue-619872748) will do for the Gallery.

Sure! Can you add it to `docs/listofexamples.rst`?

**Pull request merged**